Add healthcheck endpoints

nitisht opened the pull request:
## Description
This PR adds readiness and liveness endpoints to probe Minio server
instance health. Endpoints can only be accessed without authentication
and the paths are /minio/health/live and /minio/health/ready for
liveness and readiness respectively.

Also removes the healthcheck script used currently for Dockerfiles.
The unauthenticated healthcheck liveness endpoint is used for healthcheck
now.

## Motivation and Context
See #5357 and #5514

## How Has This Been Tested?
Manually with checking the endpoints

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [ ] All new and existing tests passed.

harshavardhana commented on cmd/healthcheck-handler.go:
```diff
@@ -0,0 +1,64 @@
+	}
+	// List buckets is unsuccessful, means server is having issues, send 503 service unavailable
+	if _, err := objLayer.ListBuckets(); err != nil {
+		writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
```
you have to return right here otherwise you will write two responses on clients.

harshavardhana commented on cmd/healthcheck-handler.go:
```diff
@@ -0,0 +1,64 @@
+	objLayer := newObjectLayerFn()
+	// Service not initialized yet
+	if objLayer == nil {
+		writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
```
you have to return right here otherwise you will write two responses on clients.

harshavardhana commented on cmd/healthcheck-handler.go:
```diff
@@ -0,0 +1,64 @@
+// that they are not ready do not receive traffic through Kubernetes Services.
+func ReadinessCheckHandler(w http.ResponseWriter, r *http.Request) {
+	if err := goroutineCountCheck(minioHealthGoroutineThreshold); err != nil {
+		writeResponse(w, http.StatusServiceUnavailable, nil, mimeNone)
```
you have to return right here otherwise you will write two responses on clients.

harshavardhana commented on cmd/healthcheck-handler.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+const (
+	minioHealthGoroutineThreshold = 500
```
500 go-routines might be really small @nitisht i would suggest we move this value to `MaxOpenFileLimit-1000`

harshavardhana reviewed (dismissed):
LGTM

nitisht committed:
Add healthcheck endpoints

This PR adds readiness and liveness endpoints to probe Minio server
instance health. Endpoints can only be accessed without authentication
and the paths are /minio/health/live and /minio/health/ready for
liveness and readiness respectively.

The new healthcheck liveness endpoint is used for Docker healthcheck
now.

Fixes #5357
Fixes #5514

harshavardhana reviewed (approved):
LGTM

nitisht committed:
Merge branch 'master' into healthcheck
